File: 0046-permutations/0046-permutations.java
class Solution {
    void dfs(int[] nums, List<List<Integer>> sol, ArrayList<Integer> cur){
        if(cur.size() == nums.length){
            sol.add(new ArrayList<Integer>(cur));
            return;
        }
        
        for(int i : nums){
            if(cur.contains(i))
                continue;
            cur.add(i);
            dfs(nums, sol, cur);
            cur.remove(cur.size()-1);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> sol = new ArrayList<>();
        dfs(nums, sol, new ArrayList<Integer>());
        return sol;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void dfs(int idx, vector<int>& nums, vector<vector<int>>& sol){
        if(idx==nums.size()){
            sol.push_back(nums);
            return;
        }
        
        for(int i=idx; i<nums.size(); i++){
            swap(nums[i], nums[idx]);
            dfs(idx+1, nums, sol);
            swap(nums[i], nums[idx]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> sol;
        dfs(0, nums, sol);
        return sol;
    }
};